Remove CacheAndSeed, utilize dataVec for similar funcitonality

IllusiveAldebaran opened the pull request:
This removes SeedVec from SearchAllSeedsDoALL's inner loop.

IllusiveAldebaran committed:
Remove seedVec from searchAllSeedsDoAll

Small refactor to SeedAlignerSearchData to support what SeedAndCache
(seedVec) used to support.

The char sequence pointer may be removable, but causes some issues at
the moment. Some commented out draft for it is there

aligner_cache.h QKey struct change to use 32->8bits for length. Same
output as before, and this is what CacheAndSeed use for length.

IllusiveAldebaran committed:
Rm CacheAndSeed, Implement get_c for SeedAlignerSearchData

TODO: Still one use of seq in startSearchSeedBi. Implement ftabSeqToInt
with sequence, not a sequence pointer

Changed QKey seq back to 32 bits (But functionally correct in both)

IllusiveAldebaran committed:
Remove char* seq from SeedAlignerSearchData

Change ftabSeqToInt into a bit computaiton

sfiligoi commented on aligner_seed.cpp:
```diff
@@ -749,43 +718,49 @@ inline bool startSearchSeedBi(
+	SeedAlignerSearchState &sstate)
 {
 
+	BwtTopBotFw bwt = sdata.bwt
```
This should not work.
bwt is changed in the function, and the changes are supposed to be propagated back to the caller.

sfiligoi commented on aligner_seed.cpp:
```diff
@@ -749,43 +718,49 @@ inline bool startSearchSeedBi(
+		if (ftabLen > 1 && ftabLen <= sdata.maxjump()) {
+
+			// Section below is equivalent to Ebwt::ftabSeqToInt(ftabLen, true, seq, off - ftabLen + 1, false);
+			uint64_t low_bits = sdata.sak.seq >> (2 * (sdata.sak.len - 1 - off));
```
Could you implement this as a function, instead of inlining?
E.g. you could call it
Ebwt::ftabSakToInt

sfiligoi commented on aligner_seed.cpp:
```diff
@@ -77,50 +77,55 @@ Constraint Constraint::editBased(int edits) {
+	, sak()
+	, nlex(0)
+	, need_reporting(false)
+	{} // may need to initialize sak with NULL and 0
```
You can remove the comment;
sak() initializer already does the proper initialization

sfiligoi commented on aligner_seed.cpp:
```diff
@@ -77,50 +77,55 @@ Constraint Constraint::editBased(int edits) {
+	// Thus c needs to be computed with bit operations
+	// 
+	int8_t get_c(int off) const {
+		int c = (sak.seq >> (2 * (sak.len-1 - off))) & 0x03;
```
You should keep the same type as you pass out, so the compiler does not have to do a double type transform.

BTW: I would recommend using "uint8_t" since it semantics says it can never be negative.

sfiligoi reviewed (changes requested):
The semantics of startSearchSeedBi changed, as you are not passing out the changes to bwt.

IllusiveAldebaran committed:
Add function ftabSakToInt()

Character type change into uint8_t
Change 'bwt' to use alias instead of new instatiation

IllusiveAldebaran commented:
Addressed these comments in a new commit.

The new ` Ebwt::ftabSakToInt()` function assumes order of the sequence and is not reversed.

sfiligoi commented on bt2_idx.h:
```diff
@@ -1416,6 +1417,29 @@ class Ebwt {
+	 */
+	static TIndexOffU ftabSakToInt(
+		const int fc,
+		const SAKey sak,
```
Sorry to be picky, but could you pass directly 
`sak_seq` and `saq_len`, 
since they are just integers,
so we do not have to include aligner_cache.h header?

sfiligoi replied:
Thanks

IllusiveAldebaran committed:
Change ftabSakToInt() seq,len parameters

sfiligoi reviewed (approved):
Looks good, thanks.
